api/logic/modplatform/atlauncher/ATLPackManifest.cpp

```diff
@@ -0,0 +1,136 @@
+        return ATLauncher::ModType::Root;
+    }
+    else if(rawType == QString("forge")) {
+        return ATLauncher::ModType::Forge;
+    }
+    else if(rawType == QString("jar")) {
+        return ATLauncher::ModType::Jar;
+    }
+    else if(rawType == QString("mods")) {
+        return ATLauncher::ModType::Mods;
+    }
+    else if(rawType == QString("flan")) {
+        return ATLauncher::ModType::Flan;
+    }
+    else if(rawType == QString("dependency")) {
+        return ATLauncher::ModType::Dependency;
+    }
+    else if(rawType == QString("ic2lib")) {
+        return ATLauncher::ModType::Ic2Lib;
+    }
+    else if(rawType == QString("denlib")) {
+        return ATLauncher::ModType::DenLib;
+    }
+    else if(rawType == QString("coremods")) {
+        return ATLauncher::ModType::Coremods;
+    }
+    else if(rawType == QString("mcpc")) {
+        return ATLauncher::ModType::MCPC;
+    }
+    else if(rawType == QString("plugins")) {
+        return ATLauncher::ModType::Plugins;
+    }
+    else if(rawType == QString("extract")) {
+        return ATLauncher::ModType::Extract;
+    }
+    else if(rawType == QString("decomp")) {
+        return ATLauncher::ModType::Decomp;
+    }
+    else if(rawType == QString("resourcepack")) {
+        return ATLauncher::ModType::ResourcePack;
+    }
+
+    return ATLauncher::ModType::Unknown;
+}
+
+static void loadVersionPack(ATLauncher::VersionPack & p, const QDomElement& ele) {
+    p.version = ele.firstChildElement("version").text();
+    p.minecraft = ele.firstChildElement("minecraft").text();
+    p.noConfigs = ele.firstChildElement("noconfigs").text() == QString("true");
+    p.mainClass = ele.firstChildElement("mainclass").text();
+    p.extraArguments = ele.firstChildElement("extraarguments").text();
+}
+
+static void loadVersionLoader(ATLauncher::VersionLoader & p, const QDomElement& ele) {
+    p.type = ele.attribute("type");
+    p.version = ele.attribute("version");
+    p.latest = ele.attribute("latest") == QString("true");
+    p.recommended = ele.attribute("recommended") == QString("true");
+    p.choose = ele.attribute("choose") == QString("true");
+}
+
+static void loadVersionLibrary(ATLauncher::VersionLibrary & p, const QDomElement& ele) {
+    p.url = ele.attribute("url");
+    p.file = ele.attribute("file");
+    p.md5 = ele.attribute("md5");
+
+    p.download_raw = ele.attribute("download");
+    p.download = parseDownloadType(p.download_raw);
+}
+
+static void loadVersionMod(ATLauncher::VersionMod & p, const QDomElement& ele) {
+    p.name = ele.attribute("name");
+    p.version = ele.attribute("version");
+    p.url = ele.attribute("url");
+    p.file = ele.attribute("file");
+    p.md5 = ele.attribute("md5");
+
+    p.download_raw = ele.attribute("download");
+    p.download = parseDownloadType(p.download_raw);
+
+    p.type_raw = ele.attribute("type");
+    p.type = parseModType(p.type_raw);
+    p.extractTo_raw = ele.attribute("extractto");
```

Comment:
Note that here we can have an `extractfolder` property which will extract the value of the `extractfolder` property from the zip file.

For instance say I have a zip with the below structure:

- /
  - somemod/
    - mod-folder/
      - modFile.txt

By not specifying `extractfolder` and if I had `extractto="root"` then I'd have a folder in my root directory called `somemod`. But by providing `extractfolder="somemod%s%sub-folder"` then the `modFile.txt` would end up in the root directory.

Note the `%s%` is used to denote a file path separator

This was mainly used for Millenaire back in the day, but is still a valid option to have.

Reply:
This should be resolved by https://github.com/MultiMC/MultiMC5/pull/3314/commits/09441ecbdfc67b743eda3444846115e79fa62c2b.